color/src/colorspace.rs

```diff
@@ -1139,10 +1139,81 @@ impl ColorSpace for Hwb {
 
 #[cfg(test)]
 mod tests {
-    use crate::{A98Rgb, ColorSpace, OpaqueColor, ProphotoRgb, Rec2020, Srgb};
+    use crate::{
+        A98Rgb, ColorSpace, DisplayP3, LinearSrgb, Oklab, OpaqueColor, ProphotoRgb, Rec2020, Srgb,
+        XyzD50, XyzD65,
+    };
 
-    fn almost_equal<CS: ColorSpace>(col1: [f32; 3], col2: [f32; 3]) -> bool {
-        OpaqueColor::<CS>::new(col1).difference(OpaqueColor::new(col2)) < 1e-4
+    fn almost_equal<CS: ColorSpace>(col1: [f32; 3], col2: [f32; 3], absolute_epsilon: f32) -> bool {
+        OpaqueColor::<CS>::new(col1).difference(OpaqueColor::new(col2)) <= absolute_epsilon
+    }
+
+    /// The maximal magnitude of the color components. Useful for calculating relative errors.
+    fn magnitude(col: [f32; 3]) -> f32 {
+        col[0].abs().max(col[1].abs()).max(col[2].abs())
+    }
+
+    #[test]
+    fn roundtrip() {
+        fn test_roundtrips<Source: ColorSpace, Dest: ColorSpace>(colors: &[[f32; 3]]) {
+            /// A tight bound on relative numerical precision.
+            const RELATIVE_EPSILON: f32 = f32::EPSILON * 16.;
+
+            for color in colors {
+                let intermediate = Source::convert::<Dest>(*color);
+                let roundtripped = Dest::convert::<Source>(intermediate);
+
+                // The roundtrip error is measured in linear sRGB. This adds more conversions, but
+                // makes the components analogous.
+                let linsrgb_color = Source::to_linear_srgb(*color);
+                let linsrgb_intermediate = Dest::to_linear_srgb(intermediate);
+                let linsrgb_roundtripped = Source::to_linear_srgb(roundtripped);
```

Comment:
It feels like these additional conversions could be dropped somehow, but I am not sure yet how. The issue is that a large magnitude in one component can cause an error relative to that magnitude in another component, but because in arbitrary color spaces the components can have quite different scales and meanings, it becomes difficult to test to a tight bound.

Reply:
Would it be feasible to have a budget of floating point numbers in between? So for example, there are 13 floats between 100 and 100.0001 (according to `100.0001f32.to_bits() - 100.0f32.to_bits()` [stackoverflow](https://stackoverflow.com/a/3588000)).

I don't know if that's exactly what you want, but just adding it as an option.

Reply:
In general that's a neat way to measure to tight bounds, I've heard that called "units in last place" before. Maybe we can use it elsewhere for our tests. I don't think it'll work for this specific test, as it's the maximal magnitude of the components that can influence the error in all components.

For example, roundtripping linear srgb (1, 0, 0) to XYZ-D65 and back using our implementation, you get an error on the order of 10^-8 in the green channel. Do the same with linear srgb (100, 0, 0) (very out-of-gamut), and you get an error on the order of 10^-6. (In this case, both errors are well below one relative float epsilon).

Reply:
Right, because it's doing big matrix multiplies. That makes sense